sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinations.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import com.google.api.services.bigquery.model.TableSchema;
+import java.io.Serializable;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.ValueInSingleWindow;
+
+/**
+ * This class provides the most general way of specifying dynamic BigQuery table destinations.
+ * Destinations can be extracted from the input element, and stored as a custom type. Mappings
+ * are provided to convert the destination into a BigQuery table reference and a BigQuery schema.
+ * The class can read a side input from another PCollection while performing these mappings.
+ *
+ * <p>For example, consider a PCollection of events, each containing a user-id field. You want to
+ * write each user's events to a separate table with a separate schema per user. Since the user-id
+ * field is a string, you will represent the destination as a string.
+ *<pre>{@code
+ *events.apply(BigQueryIO.<UserEvent>write()
+ *  .to(new DynamicDestinations<UserEvent, String>() {
```

Comment:
Is there side input support provided to the user? If yes, please document it here.